File: graphs/2127-maximum-employees-to-be-invited-to-a-meeting.go
package graphs

import "container/list"

// https://leetcode.com/problems/maximum-employees-to-be-invited-to-a-meeting/

/*
Overview:
1. Every connected component has exactly one cycle.
2. Components with a cycle of length = 2, form non-closed circle.
	They can be connected to other non closed circles.
3. Components with a cycle of length > 2, form closed circles.
	They can't be connected

Solution:
1. Find the longest cycle, aka the longest closed circle.
2. Find all components with length = 2 cycles.
	For each, find the longest path component.
	Sum all of the paths.
3. Return max of step 1 and 2.
*/

// Approach: Topological Sort to Reduce Non-Cyclic Nodes
// Time: O(n)
// Space: O(n)
func maximumInvitations(favorite []int) int {
	var n = len(favorite)
	var indegree = make([]int, n)
	for empl := 0; empl < n; empl++ {
		indegree[favorite[empl]]++
	}
	var queue = list.New()
	for empl := 0; empl < n; empl++ {
		if indegree[empl] == 0 {
			queue.PushBack(empl)
		}
	}
	var depth = make([]int, n)
	var currEmpl, nextEmpl int
	for queue.Len() > 0 {
		currEmpl = queue.Remove(queue.Front()).(int)
		nextEmpl = favorite[currEmpl]
		depth[nextEmpl] = max(depth[nextEmpl], depth[currEmpl]+1)
		indegree[nextEmpl]--
		if indegree[nextEmpl] == 0 {
			queue.PushBack(nextEmpl)
		}
	}
	var longestCycle = 0
	var twoLenCycle = 0
	var cycleLength, tmpEmpl int
	for empl := 0; empl < n; empl++ {
		if indegree[empl] == 0 {
			continue
		}
		cycleLength = 0
		tmpEmpl = empl
		for indegree[tmpEmpl] != 0 {
			indegree[tmpEmpl]--
			cycleLength++
			tmpEmpl = favorite[tmpEmpl]
		}
		if cycleLength == 2 {
			twoLenCycle += depth[empl] + depth[favorite[empl]] + 2
		} else {
			longestCycle = max(longestCycle, cycleLength)
		}
	}
	return max(longestCycle, twoLenCycle)
}

File: graphs/2127-maximum-employees-to-be-invited-to-a-meeting.py
from collections import deque
from typing import List
import unittest

# https://leetcode.com/problems/maximum-employees-to-be-invited-to-a-meeting/

# python3 -m unittest graphs/2127-maximum-employees-to-be-invited-to-a-meeting.py


class Solution(unittest.TestCase):
    # Approach: Topological Sort to Reduce Non-Cyclic Nodes
    # Time: O(n)
    # Space: O(n)
    def maximumInvitations(self, favorite: List[int]) -> int:
        n = len(favorite)
        indegree = [0] * n
        for empl in range(n):
            indegree[favorite[empl]] += 1
        queue = deque()
        for empl in range(n):
            if indegree[empl] == 0:
                queue.append(empl)
        depth = [1] * n
        while queue:
            curr_empl = queue.popleft()
            next_empl = favorite[curr_empl]
            depth[next_empl] = max(depth[next_empl], depth[curr_empl] + 1)
            indegree[next_empl] -= 1
            if indegree[next_empl] == 0:
                queue.append(next_empl)
        longest_cycle = 0
        two_len_cycle = 0
        for empl in range(n):
            if indegree[empl] == 0:
                continue
            cycle_length = 0
            tmp_empl = empl
            while indegree[tmp_empl] != 0:
                indegree[tmp_empl] -= 1
                cycle_length += 1
                tmp_empl = favorite[tmp_empl]
            if cycle_length == 2:
                two_len_cycle += depth[empl] + depth[favorite[empl]]
            else:
                longest_cycle = max(longest_cycle, cycle_length)
        return max(longest_cycle, two_len_cycle)

    def test(self):
        for favorite, expected in [
            ([2, 2, 1, 2], 3),
            ([1, 2, 0], 3),
            ([3, 0, 1, 4, 1], 4),
        ]:
            output = self.maximumInvitations(favorite)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
